src/csharp/Microsoft.Spark/ChunkedStream.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Spark.Interop.Ipc;
+
+namespace Microsoft.Spark
+{
+    public class ChunkedStream
+    {
+        private readonly int _bufferSize;
+        private byte[] _buffer;
+        private int _currentPos;
+        private Stream _wrapped;
+
+        internal ChunkedStream(Stream wrapped, int bufferSize)
+        {
+            _bufferSize = bufferSize;
+            _buffer = new byte[_bufferSize];
+            _currentPos = 0;
+            _wrapped = wrapped;
+        }
+
+        internal void WriteInt(int value, Stream stream)
+        {
+            byte[] bytes = BitConverter.GetBytes(value);
+            if (BitConverter.IsLittleEndian)
+                Array.Reverse(bytes);
+            stream.Write(bytes, 0, bytes.Length);
+        }
+
+        internal byte[] ConvertToByteArray(object value)
+        {
+            var formatter = new BinaryFormatter();
+            using (var ms = new MemoryStream())
+            {
+                formatter.Serialize(ms, value);
+                return ms.ToArray();
+            }
+        }
+
+        public void Write(object value)
```

Comment:
The flow in Python is this gets called as soon as you step into [`pickle.dump`](https://github.com/apache/spark/blob/master/python/pyspark/broadcast.py#L107) where the value being passed is the broadcast value. But the value (bytes) that [this function](https://github.com/apache/spark/blob/master/python/pyspark/serializers.py#L601) receives has a different format/padding to it. For example:

broadcast value -> bytes written to Stream

hello -> b'\x80\x02X\x05\x00\x00\x00helloq\x00.'
50 -> b'\x80\x02K2.'
foobar -> b'\x80\x02X\x06\x00\x00\x00foobarq\x00.'
hi there -> b'\x80\x02X\x08\x00\x00\x00hi thereq\x00.'
[hi there, hello, hows it going] -> b'\x80\x02]q\x00(X\x08\x00\x00\x00hi thereq\x01X\x05\x00\x00\x00helloq\x02X\x0f\x00\x00\x00how is it goingq\x03e.'

I am wondering if I need to derive a pattern from this, or if there is some hidden code that is being hit that I am missing.